chainimport/headers_import.go

```diff
@@ -0,0 +1,1919 @@
+
+	log.Infof("Processing %d headers (block and filter) beyond import range "+
+		"(heights %d-%d)", region.End-region.Start+1, region.Start, region.End)
+
+	switch options.BeyondImportRangeMode {
+	case PreserveBeyond:
+		// Nothing to do, just log.
+		log.Infof("Preserving headers (block and filter) beyond import range "+
+			"due to %s mode", options.BeyondImportRangeMode)
+
+	case TruncateBeyond:
+		// Truncate headers beyond import range.
+		if err := s.truncateHeaders(region.Start, options); err != nil {
+			return fmt.Errorf("failed to truncate headers beyond import "+
+				"range: %w", err)
+		}
+		log.Infof("Truncated all headers (block and filter) beyond height %d",
+			region.Start-1)
+	}
+
+	return nil
+}
+
+func (s *HeadersImport) processNewHeadersRegion(region HeaderRegion,
+	options ImportOptions, result *ImportResult) error {
+
+	if !region.Exists {
+		return nil
+	}
+
+	log.Infof("Adding %d new headers (block and filter) from heights %d to %d",
+		region.End-region.Start+1, region.Start, region.End)
+
+	err := s.appendNewHeaders(region.Start, region.End, options)
+	if err != nil {
+		return fmt.Errorf("failed to append new headers: %w", err)
+	}
+
+	result.ProcessedCount += int(region.End - region.Start + 1)
+	result.AddedCount += int(region.End - region.Start + 1)
+
+	return nil
+}
+
+// validateHeadersExactMatch performs exact validation of all headers in the
+// specified range, ensuring that headers from the import source match exactly
+// with their counterparts in the target store. This is used by
+// ValidateAndAppend mode to verify consistency of overlapping headers before
+// proceeding with the import operation.
+//
+// The validation is performed concurrently for improved performance since this
+// is primarily an I/O bound operation. Each worker safely validates headers at
+// different heights without interfering with other workers. The function adapts
+// to available system resources and uses a fail-fast approach, stopping all
+// validation work as soon as any mismatch is detected.
+func (s *HeadersImport) validateHeadersExactMatch(ctx context.Context,
+	startHeight, endHeight uint32, options ImportOptions) error {
+
+	// Calculate the range size and determine concurrency level.
+	rangeSize := endHeight - startHeight + 1
+
+	// If range is small, process sequentially.
+	if rangeSize <= 100 {
+		return s.validateHeadersSequential(startHeight, endHeight, options)
+	}
+
+	// Set up concurrency level based on CPU cores, but cap reasonably.
+	numWorkers := min(runtime.NumCPU(), 8)
+
+	// Create channels for work distribution and error collection.
+	heightChan := make(chan uint32, rangeSize)
+	errChan := make(chan error, rangeSize)
+	doneChan := make(chan struct{})
+	var wg sync.WaitGroup
+
+	// Create a derived context that can be canceled either by parent or local
+	// errors
+	validationCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Start workers.
+	for range numWorkers {
+		wg.Add(1)
+		go func() {
```

Comment:
This whole goroutine should be extracted into a method to avoid too much indentation.